src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,151 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory
+{
+    [DataContract]
+    internal sealed class InstanceDiscoveryResponse
+    {
+        [DataMember(Name = "tenant_discovery_endpoint")]
+        public string TenantDiscoveryEndpoint { get; set; }
+
+        [DataMember(Name = "metadata")]
+        public InstanceDiscoveryMetadataEntry[] Metadata { get; set; }
+    }
+
+    [DataContract]
+    internal sealed class InstanceDiscoveryMetadataEntry
+    {
+        [DataMember(Name = "preferred_network")]
+        public string PreferredNetwork { get; set; }
+
+        [DataMember(Name = "preferred_cache")]
+        public string PreferredCache { get; set; }
+
+        [DataMember(Name = "aliases")]
+        public string[] Aliases { get; set; }
+    }
+
+    internal class InstanceDiscovery
+    {
+        public static string DefaultTrustedAuthority = "login.microsoftonline.com";
+
+        private static HashSet<string> WhitelistedAuthorities = new HashSet<string>(new[]
+        {
+            "login.windows.net", // Microsoft Azure Worldwide - Used in validation scenarios where host is not this list 
+            "login.chinacloudapi.cn", // Microsoft Azure China
+            "login.microsoftonline.de", // Microsoft Azure Blackforest
+            "login-us.microsoftonline.com", // Microsoft Azure US Government - Legacy
+            "login.microsoftonline.us", // Microsoft Azure US Government
+            "login.microsoftonline.com" // Microsoft Azure Worldwide
+        });
+
+        public const string AuthorizeEndpointTemplate = "https://{host}/{tenant}/oauth2/authorize";
+
+        // The following cache could be private, but we keep it public so that internal unit test can take a peek into it
+        public static ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry> InstanceCache =
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Change the visibility of 'InstanceCache' or make it 'const' or 'readonly'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS2223)